File: core/src/main/java/com/lomekwi/cave/pipeline/audio/AudClipSrc.java
package com.lomekwi.cave.pipeline.audio;

import com.badlogic.gdx.scenes.scene2d.Actor;
import com.lomekwi.cave.app.AppAudioOut;
import com.lomekwi.cave.pipeline.Frame;
import com.lomekwi.cave.pipeline.Source;
import com.lomekwi.cave.resource.media.AudRes;
import com.lomekwi.cave.timeline.Track;
import com.lomekwi.cave.ui.editpanel.detail.AudClipSrcActor;

import java.io.Serial;

public class AudClipSrc extends Source<AudFrame> {
    private AudRes audRes;
    @Serial
    private static final long serialVersionUID = 1L;

    public AudClipSrc(AudRes audRes) {
        super();
        this.audRes = audRes;
    }

    public AudRes getAudRes() {
        return audRes;
    }

    @Override
    public void sync(long time, Track track) throws Exception {
        audRes.sync(track.index, time);
    }

    @Override
    protected AudFrame generate(long time, Track track) {

        if (frame == null || frame.track != track) {
            frame = new AudFrame(AppAudioOut.SAMPLE_RATE, 2, track, this);
        }

        try {
            audRes.get(track.index, time, frame);
            if (frame.getSamples() == null) return null;
            return frame;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public long getLengthPerExportFrame() {
        return audRes.getFrameLength();
    }
    @Override
    public long getDuration() {
        return audRes.getDuration();
    }
    @Override
    public Class<? extends Frame> getFrameType() {
        return AudFrame.class;
    }
    @Override
    public String getDisplayName() {
        return "音频源";
    }
    @Override
    public void onDuplicate(Source<?> original) {
        AudClipSrc src = (AudClipSrc) original;
        this.audRes = src.audRes;
    }
    @Override
    public Actor getDetailActor() {
        return new AudClipSrcActor(this);
    }
}

File: core/src/main/scala/com/lomekwi/cave/pipeline/audio/AudClipSrc.scala
package com.lomekwi.cave.pipeline.audio

import com.lomekwi.cave.app.AppAudioOut
import com.lomekwi.cave.pipeline.Node
import com.lomekwi.cave.pipeline.Source
import com.lomekwi.cave.pipeline.num.NumFrame
import com.lomekwi.cave.resource.media.AudRes
import com.lomekwi.cave.timeline.Track
import com.lomekwi.cave.ui.editpanel.tlarea.TlAudSrcActor
import com.lomekwi.cave.ui.editpanel.tlarea.TlSrcActor

@SerialVersionUID(1L)
class AudClipSrc(private var audRes: AudRes) extends Source[AudFrame] {
  addOutPort(new Node.OutPort[NumFrame]("时长", classOf[NumFrame]) {
    private final val `val`: NumFrame = new NumFrame(null)

    override def getData: NumFrame = {
      `val`.setVal(getDuration.toDouble)
      `val`
    }
  })

  def getAudRes: AudRes = {
    audRes
  }

  override def sync(time: Long, track: Track): Unit = {
    audRes.sync(track.index, time)
  }

  override protected def generate(time: Long, track: Track): AudFrame = {

    if (frame == null || (frame.track ne track)) {
      frame = new AudFrame(AppAudioOut.SAMPLE_RATE, track, this)
    }

    try {
      audRes.get(track.index, time, frame)
      if (frame.getSamples == null) null else frame
    } catch {
      case e: Exception =>
        throw new RuntimeException(e)
    }
  }

  override def getLengthPerExportFrame: Long = {
    audRes.getFrameLength
  }
  override def getDuration: Long = {
    audRes.getDuration
  }
  override def getFrameType: Class[AudFrame] = {
    classOf[AudFrame]
  }
  override def getDisplayName: String = {
    "音频源"
  }
  override def onDuplicate(original: Source[?]): Unit = {
    val src = original.asInstanceOf[AudClipSrc]
    this.audRes = src.audRes
  }
  override def createTlSrcActor(): TlSrcActor = {
    new TlAudSrcActor(this)
  }
}
